client/src/document/organisms/sidebar/index.tsx

```diff
@@ -68,13 +69,148 @@ export function SidebarContainer({
   );
 }
 
+function getLineDistance(a: HTMLElement | null, b: HTMLElement | null): number {
+  if (!a || !b) {
+    return -1;
+  }
+
+  const { top: aTop, bottom: aBottom } = a.getBoundingClientRect();
+  const { top: bTop, bottom: bBottom } = b.getBoundingClientRect();
+
+  const px = aTop < bTop ? bBottom - aTop : aBottom - bTop;
+  const rem = px / 16;
+
+  return Math.round(rem);
+}
+
+function getTreePath(
+  element: HTMLElement,
+  { boundary, selector }: { boundary: HTMLElement; selector: string }
+): HTMLElement[] {
+  const path: HTMLElement[] = [];
+
+  let current: HTMLElement | null = element;
+  while (current && boundary.contains(current)) {
+    path.push(current);
+    current = (current.parentNode as HTMLElement)?.closest(selector);
+  }
+
+  return path.reverse();
+}
+
+function getTreeDistance(
+  a: HTMLElement | null,
+  b: HTMLElement | null,
+  { boundary, selector }: { boundary: HTMLElement; selector: string }
+): number {
+  if (!a || !b) {
+    return -1;
+  }
+  if (a === b) {
+    return 0;
+  }
+
+  const aPath = getTreePath(a, { boundary, selector });
+  const bPath = getTreePath(b, { boundary, selector });
+
+  while (aPath.length && bPath.length && aPath[0] === bPath[0]) {
+    // Remove common ancestors.
+    aPath.shift();
+    bPath.shift();
+  }
+
+  // Remove one edge.
+  aPath.pop() || bPath.pop();
+
+  return aPath.length + bPath.length;
+}
+
+function getSlugPath(slug: string): string[] {
+  return slug.split("/");
+}
+
+function getSlugDistance(a: string | null, b: string | null) {
+  if (!a || !b) {
+    return -1;
+  } else if (a === b) {
+    return 0;
+  }
+
+  const aPath = getSlugPath(a);
+  const bPath = getSlugPath(b);
+
+  while (aPath.length && bPath.length && aPath[0] === bPath[0]) {
+    // Remove common ancestors.
+    aPath.shift();
+    bPath.shift();
+  }
+
+  // Remove one edge.
+  aPath.pop() || bPath.pop();
+
+  return aPath.length + bPath.length;
```

Comment:
Extract to separate function.

```suggestion
  while (aPath.length && bPath.length && aPath[0] === bPath[0]) {
    // Remove common ancestors.
    aPath.shift();
    bPath.shift();
  }

  // Remove one edge.
  aPath.pop() || bPath.pop();

  return aPath.length + bPath.length;
```